usr/lib/pkcs11/common/trace.c

```diff
@@ -206,41 +206,69 @@ CK_RV trace_initialize(void)
     return (CKR_FUNCTION_FAILED);
 }
 
-void ock_traceit(trace_level_t level, const char *fmt, ...)
+void ock_traceit(trace_level_t level, const char *file, int line,
+                 const char *stdll_name, const char *fmt, ...)
 {
     va_list ap;
     time_t t;
     struct tm *tm;
+    const char *fmt_pre;
     char buf[1024];
     char *pbuf;
     int buflen, len;
 
     if (trace.fd < 0)
         return;
 
-    if (level <= trace.level) {
-        pbuf = buf;
-        buflen = sizeof(buf);
-
-        /* add the current time */
-        t = time(0);
-        tm = localtime(&t);
-        len = strftime(pbuf, buflen, "%m/%d/%Y %H:%M:%S ", tm);
-        pbuf += len;
-        buflen -= len;
-        /* add the current time */
-
-        /* add the format */
-        va_start(ap, fmt);
-        vsnprintf(pbuf, buflen, fmt, ap);
-        va_end(ap);
-
-        /* serialize appends to the file */
-        pthread_mutex_lock(&tlmtx);
-        if (write(trace.fd, buf, strlen(buf)) == -1)
-            fprintf(stderr, "cannot write to trace file\n");
-        pthread_mutex_unlock(&tlmtx);
+    if (level > trace.level)
+        return;
+
+    pbuf = buf;
+    buflen = sizeof(buf);
+
+    /* add the current time */
+    t = time(0);
+    tm = localtime(&t);
+    len = strftime(pbuf, buflen, "%m/%d/%Y %H:%M:%S ", tm);
+    pbuf += len;
+    buflen -= len;
+
+    /* add file line and stdll name */
+    switch (level) {
+    case TRACE_LEVEL_ERROR:
+        fmt_pre = "[%s:%d %s] ERROR: ";
+        break;
+    case TRACE_LEVEL_WARNING:
+        fmt_pre = "[%s:%d %s] WARN: ";
+        break;
+    case TRACE_LEVEL_INFO:
+        fmt_pre = "[%s:%d %s] INFO: ";
+        break;
+    case TRACE_LEVEL_DEVEL:
+        fmt_pre = "[%s:%d %s] DEVEL: ";
+        break;
+    case TRACE_LEVEL_DEBUG:
+        fmt_pre = "[%s:%d %s] DEBUG: ";
+        break;
+    default:	/* cannot happen */
+        fmt_pre = "[%s:%d %s] ERROR: ";
+        break;
     }
+    snprintf(pbuf, buflen, fmt_pre, file, line, stdll_name);
+
+    /* add the format */
+    len = strlen(buf);
+    pbuf = buf + len;
+    buflen = sizeof(buf) - len;
+    va_start(ap, fmt);
+    vsnprintf(pbuf, buflen, fmt, ap);
+    va_end(ap);
+
+    /* serialize appends to the file */
+    pthread_mutex_lock(&tlmtx);
+    if (write(trace.fd, buf, strlen(buf)) == -1)
+        fprintf(stderr, "cannot write to trace file\n");
```

Comment:
This is a bad implementation: write may return with > 0 but < strlen(buf).
However, much better would be to use the stdio functions fopen/fwrite/fclose instead of the bare system calls. For example there is no mutex needed then.

Reply:
has been working forever like that. will not be fixed with this warning-related PR.